service/storage_proxy.cc

```diff
@@ -1063,17 +1075,17 @@ class hint_mutation : public shared_mutation {
     }
     virtual future<> apply_locally(storage_proxy& sp, storage_proxy::clock_type::time_point timeout,
             tracing::trace_state_ptr tr_state, db::per_partition_rate_limit::info rate_limit_info,
-            fencing_token) override {
+            fencing_token fence) override {
         // A hint will be sent to all relevant endpoints when the endpoint it was originally intended for
         // becomes unavailable - this might include the current node
-        return sp.mutate_hint(_schema, *_mutation, std::move(tr_state), timeout);
+        return sp.apply_fence(sp.mutate_hint(_schema, *_mutation, std::move(tr_state), timeout), fence, utils::fb_utilities::get_broadcast_address());
```

Comment:
Why are we applying the fence after the damage was done?

Reply:
Since this is part of the coordinator-side request, do we need apply_fence() at all? All coordinators are drained by the barrier, so it should be impossible for this apply_fence() to fail.  

@avikivity The apply_fence() logic matches how replica-side fencing works in general, for regular requests as well. It doesn't wait for replica-side requests, only makes sure that they fail. This is fine now, because the barrier is not said to drain replicas. But we will need to change it later when we implement cleanup, which needs to drain replicas, for both vnodes and tablets. @kostja @gleb-cloudius What was your plan for that?

Reply:
Ok. I don't like it, but let's not change the mechanism while we implement it.

In the future I'd like drain to work as a wave: first drain the super-coordinator, then the coordiantor, then counter-leaders, then replicas, then view replicas.

btw, long-running requests should be able to listen to drain requests and switch to a new topology during a request, or perhaps abort + retry, so they don't induce extra latency.

Reply:
> But we will need to change it later when we implement cleanup, which needs to drain replicas, for both vnodes and tablets. @kostja @gleb-cloudius What was your plan for that?

The plan is described in https://docs.google.com/document/d/1X6pARlxOy6KRQ32JN8yiGsnWA9Dwqnhtk7kMDo8m9pI/. Cleanup will be uncoordinated (the decision that a node should do a cleanup will be done by the coordinator though). A node that need to do a cleanup will do local drain before cleanup. The reason to do it like that we already discussed on the first fencing patch IIRC. To summarize since the cleanup should be done as part of a recovery from failed operation it should be as resilient to failures as possible and since the minimum we require for topology operation to proceed (either to completion or abortion) is a presence of raft quorum making cleanup operation to only set a flag in node's state fulfill the resilience requirement.

Reply:
> Since this is part of the coordinator-side request, do we need apply_fence() at all? All coordinators are drained by the barrier, so it should be impossible for this apply_fence() to fail.

There was this dialog in the initial fencing PR:
[gleb-cloudius](https://github.com/gleb-cloudius) [on May 16](https://github.com/scylladb/scylladb/pull/14580#r1195212684)
Shouldn't we also do fencing on local writes? If we do drain with timeout a local write may complete after the fence moved.
 
Member
Author
@[gusev-p](https://github.com/gusev-p) gusev-p [on May 16](https://github.com/scylladb/scylladb/pull/14580#r1195283907)
I missed lmutate in send_to_live_endpoints, will fix. On replicas try_apply_fence is done in handle_write.

I think I've misapprehended the situation here. `send_to_live_endpoints/lmutate/apply_locally` are called only on coordinators, and if the drain timeout happens with should rollback the topology change operation, so there is no need in fencing in `apply_locally`? @gleb-cloudius

Reply:
> > Since this is part of the coordinator-side request, do we need apply_fence() at all? All coordinators are drained by the barrier, so it should be impossible for this apply_fence() to fail.
> 
> Topology change coordinator drains all nodes, it can get timeout from some of them, whether this timeout causes topology change op to abort depends on whether this node is affected or not. If coordinator was also a replica for some data that is not part of the affected ranges it can still be fenced.

In the current implementation of the barrier all nodes are in the affected set. Are you developing with a different implementation in mind?.

Reply:
> The plan is described in

What I meant was what is the plan for implementing the local drain, which we will need eventually. It would plug into the same places were we currently put apply_fence(), but in a different way. Maybe we don't need two mechanisms for that.

Is it necessary to fail after at the end of the operation? What invariant does it try to uphold? It seems racy, the replica-side operation could have ended just before it was fenced, but the coordinator could get the response after we fenced.

Reply:
If we don't need to fail at the end, we can just use RAII for tracking, like a gate guard.

Reply:
I'm implementing this [spec](https://docs.google.com/document/d/1X6pARlxOy6KRQ32JN8yiGsnWA9Dwqnhtk7kMDo8m9pI/edit) by @gleb-cloudius, it separates affected replicas in `global_token_metadata_barrier`, it's important for tablets IIUC. `apply_fence` in `apply_locally` was introduced in the initial fencing PR, because of [this comment](https://github.com/scylladb/scylladb/pull/13884/files#r1195212684):

@[gleb-cloudius](https://github.com/gleb-cloudius) gleb-cloudius [on May 16](https://github.com/scylladb/scylladb/pull/14580#r1195212684)
Shouldn't we also do fencing on local writes? If we do drain with timeout a local write may complete after the fence moved.

Maybe we should have skipped them and add back later when there is a clear need for them. I agree that now it's difficult to reason about the implementation, because a part of it serves the future needs.

Reply:
> Is it necessary to fail after at the end of the operation? 

Yes. If the fencing barrier completes before local write completes the write may not be included in the streaming and should be reported as failed.

> What invariant does it try to uphold?

All acked writes are included in the streaming,

> It seems racy, the replica-side operation could have ended just before it was fenced, but the coordinator could get the response after we fenced.

If operartion ends before it is fenced it means it will be included in the streaming and it is fine for the coordinator to get the positive response.